internal/app/orchestrator/orchestrator.go

```diff
@@ -90,16 +110,176 @@ func New(ctx context.Context, l log.Logger, mapper mapper.Mapper, upstreamClient
 //
 // Cancel is an optional function to release resources in the producer. If
 // provided, the consumer may call this function multiple times.
-func (c *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
-	// TODO implement.
-	return nil, nil
+func (o *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
+	ctx := context.Background()
+
+	// If this is the first time we're seeing the request from the
+	// downstream client, initialize a channel to feed future responses.
+	responseChannel := o.downstreamResponseMap.createChannel(&req)
+
+	aggregatedKey, err := o.mapper.GetKey(req)
+	if err != nil {
+		// Can't map the request to an aggregated key. Log and continue to
+		// propagate the response upstream without aggregation.
+		o.logger.With("err", err).With("req node", req.GetNode()).Warn(ctx, "failed to map to aggregated key")
+		// Mimic the aggregated key.
+		// TODO (https://github.com/envoyproxy/xds-relay/issues/56). This key
+		// needs to be made more granular to uniquely identify a request.
+		aggregatedKey = fmt.Sprintf("%s%s_%s", unaggregatedPrefix, req.GetNode().GetId(), req.GetTypeUrl())
+	}
+
+	// Register the watch for future responses.
+	err = o.cache.AddRequest(aggregatedKey, &req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req node", req.GetNode()).Error(ctx, "failed to add watch")
+		closedChannel := o.downstreamResponseMap.delete(&req)
+		return closedChannel, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Warn(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil && cached.Resp.Raw.GetVersionInfo() != req.GetVersionInfo() {
+		// If we have a cached response and the version is different,
+		// immediately push the result to the response channel.
+		go func() { responseChannel <- convertToGcpResponse(cached.Resp, req) }()
+	}
+
+	// Check if we have a upstream stream open for this aggregated key. If not,
+	// open a stream with the representative request.
+	//
+	// Locking is necessary here so that a simultaneous downstream request
+	// that maps to the same aggregated key doesn't result in two upstream
+	// streams.
+	o.upstreamResponseMap.mu.Lock()
+	if _, ok := o.upstreamResponseMap.responseChannel[aggregatedKey]; !ok {
+		upstreamResponseChan := o.upstreamClient.OpenStream(ctx, &req)
```

Comment:
i'm still a bit unclear about how to split the responsibilities between `ctx` and `done` channel.
The grpc streams are based on the `ctx` and `RecvMsg` will return when ctx is cancelled. We still need a way to cancel the context for `RecvMsg` to stop blocking and clean up the goroutine. Is there a different way you envision the stream can be cancelled?

Reply:
That is not necessarily true. You can have OpenStream take in a shutdown channel and then have it not listen to RecvMsg anymore and exit the for-loop. Ideally, we can push an EOF to envoy or a GOAWAY header so that envoy will close the connection before exiting the loop, but I don't think that is as important.

Reply:
based on [this](https://github.com/grpc/grpc-go/blob/master/stream.go#L114-L117) RecvMsg is a blocking call and will block irrespective. the only way to stop that is by [cancelling](https://github.com/grpc/grpc-go/blob/master/stream.go#L132) the context

Reply:
You can leverage `select` to listen for both RecvMsg or the cancel channel.

Reply:
I think there's some more details that are important here. The ownership of the `done` channel is with upstream. This means upstream can communicate to the orchestrator to shutdown. However i don't the ownership model be inverted like that. When orchestrator figures that a key is not important anymore or ttl expires and the upstream stream has to be released, or in any other clean up case, how does orchestrator signal that to the upstream client? I'm not necessarily suggesting to use ctx, but we need something for the usecase i mentioned.

Reply:
I think you can have two channels for this responsibility. One for the upstream to communicate that it is shutting down and the orchestrator to communicate up that the stream is no longer valid.  I am fine with you using ctx for this as well, but I am not sure how you will be able to do it without storing ctx on the datastructure. Ideally, it would be great if CreateWatch accepted a context.

Reply:
I believe we've agreed on a solution offline and this is reflected in the upstream client implementation with the returned callback, but let me know if that's not the case.